jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/BufferedContentSinkTest.java

```diff
@@ -0,0 +1,277 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.eclipse.jetty.io.content.AsyncContent;
+import org.eclipse.jetty.io.content.BufferedContentSink;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Callback;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class BufferedContentSinkTest
+{
+    private ArrayByteBufferPool.Tracking _bufferPool;
+
+    @BeforeEach
+    public void setUp()
+    {
+        _bufferPool = new ArrayByteBufferPool.Tracking();
+    }
+
+    @AfterEach
+    public void tearDown()
+    {
+        assertThat("Leaks: " + _bufferPool.dumpLeaks(), _bufferPool.getLeaks().size(), is(0));
+    }
+
+    @Test
+    public void testWrongMaxBufferSize()
+    {
+        assertThrows(IllegalArgumentException.class, () -> new BufferedContentSink(new AsyncContent(), _bufferPool, true, 0));
+        assertThrows(IllegalArgumentException.class, () -> new BufferedContentSink(new AsyncContent(), _bufferPool, true, -1));
+    }
+
```

Comment:
We need a test like:
```suggestion


    @Test
    public void testByteByByteRecursion() throws Exception
    {
        try (AsyncContent async = new AsyncContent())
        {
            BufferedContentSink buffered = new BufferedContentSink(async, _bufferPool, true, 4096);

            AtomicInteger count = new AtomicInteger(8192);
            CountDownLatch complete = new CountDownLatch(1);

            Callback callback = new Callback()
            {
                @Override
                public void succeeded()
                {
                    int c = count.decrementAndGet();
                    if (c > 0)
                        buffered.write(false, ByteBuffer.wrap(new byte[1]), this);
                    else if (c == 0)
                        buffered.write(false, ByteBuffer.wrap(new byte[1]), this);
                    else
                        complete.countDown();
                }
            };
            
            callback.succeeded();
            assertTrue(complete.await(10, TimeUnit.SECONDS));
            assertThat(count.get(), is(0));
            // TODO more checks
        }
    }
```